cloud-functions/failed_fhir_conversion/main.py

```diff
@@ -0,0 +1,57 @@
+import json
+import functions_framework
+import flask
+import os
+from google.cloud import storage
+from phdi_cloud_function_utils import (
+    check_for_environment_variables,
+    make_response,
+    validate_request_body_json,
+    validate_request_header,
+)
+
+
+@functions_framework.http
+def failed_fhir_conversion(request: flask.Request) -> flask.Response:
+    """
+    When the FHIR Converter fails to convert a message, this function is called.
+    The output JSON, including the original input data and parameters,
+    are saved to a file in a storage bucket.
+
+    :param request: A Flask POST request object. The header must contain
+        'Content-Type:application/json' and the body must be valid JSON.
+    :return: Returns a flask.Response object
+    """
+
+    content_type = "application/json"
+    # Validate request header.
+    header_response = validate_request_header(request, content_type)
+    if header_response.status_code == 400:
+        return header_response
+
+    # Validate request body.
+    body_response = validate_request_body_json(request)
+    if body_response.status_code == 400:
+        return body_response
+
+    # Check for the required environment variables.
+    environment_check_response = check_for_environment_variables(
+        ["FAILED_FHIR_CONVERSION_BUCKET"]
+    )
+    if environment_check_response.status_code == 500:
+        return environment_check_response
+
+    # Upload file to storage bucket.
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(os.environ.get("FAILED_FHIR_CONVERSION_BUCKET"))
```

Comment:
I had thought we would set `FAILED_FHIR_CONVERSION_BUCKET` in our orchestration layer and pass it as part of `request`. Going this way would:

1. Decrease some reliance on environment variables which can complicate things. (not the strongest argument)
2. It always this function to be a little more generic by not restricting it to a single bucket.

Reply:
Actually, I think there is value in only allowing this function to write to our bucket that is intended to contain PHI as it mitigates the risk that PHI/PII is written somewhere else inadvertently. I'm on board with with environment variable approach. Nice work!